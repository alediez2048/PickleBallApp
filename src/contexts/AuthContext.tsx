import React, {
  createContext,
  useContext,
  useState,
  useEffect,
  useCallback,
} from "react";
import { router } from "expo-router";
import {
  mockApi,
  FirstTimeProfileData,
  UpdateProfileData,
} from "@/services/mockApi";
import { storage } from "@/services/storage";
import { socialAuth } from "@/services/socialAuth";
import { Alert, Platform } from "react-native";
import { MembershipPlan } from "@/types/membership";
import {
  signUpWithEmail,
  signInWithEmail,
  getSession,
  logout,
} from "@/services/authService";
import { createUserProfile } from "@/services/userService";
import AsyncStorage from "@react-native-async-storage/async-storage";

interface PaymentMethod {
  id: string;
  last4: string;
  brand: string;
  expiryMonth: string;
  expiryYear: string;
  isDefault: boolean;
}

interface UserProfile {
  id?: string;
  email?: string;
  name?: string;
  isVerified?: boolean;
  skillLevel?: string;
  profileImage?:
    | string
    | {
        uri: string;
        base64: string;
        timestamp: number;
      };
  phoneNumber?: string;
  dateOfBirth?: string;
  address?: {
    street?: string;
    city?: string;
    state?: string;
    zipCode?: string;
    country?: string;
  };
  membership?: MembershipPlan;
  paymentMethods?: PaymentMethod[];
  hasCompletedProfile?: boolean;
}

interface AuthState {
  user: UserProfile | null;
  token: string | null;
  session: any | null;
  isLoading: boolean;
}

interface AuthContextType extends AuthState {
  // Authentication methods
  signIn: (email: string, password: string) => Promise<void>;
  signUp: (email: string, password: string, name: string) => Promise<void>;
  signOut: () => Promise<void>;

  // Social authentication methods
  signInWithGoogle: () => Promise<void>;
  signInWithFacebook: () => Promise<void>;

  // Profile management methods
  updateProfile: (updates: Partial<UserProfile>) => Promise<void>;
  updateFirstTimeProfile: (data: FirstTimeProfileData) => Promise<void>;
  updateMembership: (plan: MembershipPlan) => Promise<void>;
  updatePaymentMethods: (methods: PaymentMethod[]) => Promise<void>;
  updatePaymentMethod?: (hasPaymentMethod: boolean) => Promise<void>;

  // Auth state
  isAuthenticated: boolean;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

export function AuthProvider({ children }: { children: React.ReactNode }) {
  const [state, setState] = useState<AuthState>({
    user: null,
    token: null,
    isLoading: true,
  });

  useEffect(() => {
    // Load token and user data from storage on app start
    loadStoredAuth();
  }, []);

  useEffect(() => {
    // Only navigate when auth state changes, not on every loading state change
    if (!state.isLoading && state.token === null) {
      router.replace("/login");
    }
  }, [state.token]);

  const loadStoredAuth = async () => {
    try {
      console.log("Loading stored auth data...");

      const [token, userString] = await Promise.all([
        AsyncStorage.getItem("auth_token"),
        AsyncStorage.getItem("user"),
      ]);

      console.log("Stored auth data:", { token, userString });

      if (token && userString) {
        const user = JSON.parse(userString);

        const { data, error } = await getSession();

        if (error || !data?.session) {
          console.warn("Session expired or invalid in Supabase");
          await AsyncStorage.multiRemove(["auth_token", "user"]);
          setState({
            token: null,
            session: null,
            user: null,
            isLoading: false,
          });
          return;
        }

        console.log("Supabase session is valid");
        setState({
          token: data.session.access_token,
          session: data.session,
          user,
          isLoading: false,
        });
      } else {
        setState({ token: null, session: null, user: null, isLoading: false });
      }
    } catch (error) {
      console.error("Error loading auth data:", error);
      setState({ token: null, session: null, user: null, isLoading: false });
    }
  };

  const signUp = async (email: string, password: string, name: string) => {
    try {
      console.log("[Auth] Starting registration process", { email, name });
      setState((prev) => ({ ...prev, isLoading: true }));

      const { data, error } = await signUpWithEmail(email, password);
      if (error) throw new Error(error.message);

      const userId = data?.user?.id;
      if (userId) {
        await createUserProfile(userId, { name, email });
        setState({
          session: data.session,
          token: data.session?.access_token || null,
          user: data.user,
          isLoading: false,
        });

        // Store the auth data
        await Promise.all([
          data.session?.access_token
            ? storage.setItem("auth_token", data.session.access_token)
            : Promise.resolve(),
          storage.setItem("user", JSON.stringify(data.user)),
        ]);
        console.log("[Auth] User data stored in local storage");
      }
    } catch (error) {
      console.error("[Auth] Registration failed:", error);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const signIn = async (email: string, password: string) => {
    try {
      console.log("Signing in...", { email });
      setState((prev) => ({ ...prev, isLoading: true }));

      const { data, error } = await signInWithEmail(email, password);

      if (error || !data.session) {
        throw new Error(error?.message || "No session returned");
      }

      const token = data.session.access_token;
      const user = data.user;
      const session = data.session;

      console.log("Sign in successful:", { token, user });

      await Promise.all([
        AsyncStorage.setItem("auth_token", token),
        AsyncStorage.setItem("user", JSON.stringify(user)),
      ]);

      console.log("Auth data stored");

      setState({ token, session, user, isLoading: false });
    } catch (error: any) {
      console.error("Sign in error:", error.message);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const signOut = async () => {
    try {
      console.log("Signing out...");
      setState((prev) => ({ ...prev, isLoading: true }));

      await Promise.all([
        storage.removeItem("auth_token"),
        storage.removeItem("user"),
      ]);
      console.log("Auth data removed");

      setState({ token: null, user: null, isLoading: false });
    } catch (error) {
      console.error("Sign out error:", error);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const handleSocialAuthSuccess = async (
    provider: "google" | "facebook",
    token: string,
    user: { id: string; email: string; name: string; photoUrl?: string }
  ) => {
    try {
      const response = await mockApi.socialAuth({ token, provider, user });

      await Promise.all([
        storage.setItem("auth_token", response.token),
        storage.setItem("user", JSON.stringify(response.user)),
      ]);

      setState({
        token: response.token,
        user: response.user,
        isLoading: false,
      });
    } catch (error) {
      console.error(`${provider} auth error:`, error);
      Alert.alert("Error", `Failed to authenticate with ${provider}`);
    }
  };

  const signInWithGoogle = async () => {
    try {
      setState((prev) => ({ ...prev, isLoading: true }));
      const result = await socialAuth.signInWithGoogle();

      if (result.type === "success" && result.token && result.user) {
        await handleSocialAuthSuccess("google", result.token, result.user);
      } else {
        throw new Error(result.message || "Google sign in failed");
      }
    } catch (error) {
      console.error("Google sign in error:", error);
      Alert.alert("Error", "Failed to sign in with Google");
      setState((prev) => ({ ...prev, isLoading: false }));
    }
  };

  const signInWithFacebook = async () => {
    try {
      setState((prev) => ({ ...prev, isLoading: true }));
      const result = await socialAuth.signInWithFacebook();

      if (result.type === "success" && result.token && result.user) {
        await handleSocialAuthSuccess("facebook", result.token, result.user);
      } else {
        throw new Error(result.message || "Facebook sign in failed");
      }
    } catch (error) {
      console.error("Facebook sign in error:", error);
      Alert.alert("Error", "Failed to sign in with Facebook");
      setState((prev) => ({ ...prev, isLoading: false }));
    }
  };

  const updateProfile = async (updates: Partial<UserProfile>) => {
    try {
      setState((prev) => ({ ...prev, isLoading: true }));

      if (!state.user?.email) {
        throw new Error("No authenticated user");
      }

      // Map UserProfile to UpdateProfileData
      const profileUpdates: UpdateProfileData = { ...updates } as any;

      // Handle address mapping if it exists
      if (updates.address) {
        profileUpdates.address = {
          address: updates.address.street || "",
          city: updates.address.city || "",
          state: updates.address.state || "",
          zipCode: updates.address.zipCode || "",
          country: updates.address.country || "United States",
        };
      }

      const { user: updatedUser } = await mockApi.updateProfile(
        state.user.email,
        profileUpdates
      );

      // Store the updated user data
      await storage.setItem("user", JSON.stringify(updatedUser));

      setState((prev) => ({
        ...prev,
        user: updatedUser,
        isLoading: false,
      }));

      console.log("Profile updated successfully:", updatedUser);
    } catch (error) {
      console.error("Profile update error:", error);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const updateFirstTimeProfile = async (data: FirstTimeProfileData) => {
    try {
      console.debug("[AuthContext] Starting first time profile update", {
        platform: Platform.OS,
        hasUser: !!state.user,
        userEmail: state.user?.email,
        data,
      });

      setState((prev) => ({ ...prev, isLoading: true }));

      if (!state.user?.email) {
        console.error("[AuthContext] No authenticated user found");
        throw new Error("No authenticated user");
      }

      // Map FirstTimeProfileData to UpdateProfileData
      const profileData: UpdateProfileData = {
        skillLevel: data.skillLevel,
        displayName: data.displayName,
        phoneNumber: data.phoneNumber,
        dateOfBirth: data.dateOfBirth,
        address: {
          address: data.address.address,
          city: data.address.city,
          state: data.address.state,
          zipCode: data.address.zipCode,
          country: data.address.country,
        },
        hasCompletedProfile: true,
      };

      console.debug("[AuthContext] Calling mockApi.updateProfile", {
        email: state.user.email,
        profileData,
      });

      const { user: updatedUser } = await mockApi.updateProfile(
        state.user.email,
        profileData
      );

      console.debug("[AuthContext] Profile update API call successful", {
        updatedUser,
      });

      // After successful profile update, update the stored user data
      const userWithProfile = {
        ...updatedUser,
        hasCompletedProfile: true,
      };

      console.debug("[AuthContext] Storing updated user data");
      await storage.setItem("user", JSON.stringify(userWithProfile));

      console.debug("[AuthContext] Updating state with new user data");
      setState((prev) => ({
        ...prev,
        user: userWithProfile,
        isLoading: false,
      }));

      console.debug(
        "[AuthContext] First time profile update completed successfully"
      );
    } catch (error) {
      console.error("[AuthContext] Profile update error:", error);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const updateMembership = async (plan: MembershipPlan) => {
    try {
      if (!state.user) {
        throw new Error("No authenticated user");
      }

      // Update user with new membership plan
      const updatedUser = {
        ...state.user,
        membership: plan,
      };

      // Store the updated user data
      await storage.setItem("user", JSON.stringify(updatedUser));

      // Update state
      setState((prev) => ({
        ...prev,
        user: updatedUser,
      }));

      console.log("Membership updated successfully:", plan);
    } catch (error) {
      console.error("Error updating membership:", error);
      throw error;
    }
  };

  const updatePaymentMethods = async (methods: PaymentMethod[]) => {
    try {
      if (!state.user) {
        throw new Error("No authenticated user");
      }

      // Update user with new payment methods
      const updatedUser = {
        ...state.user,
        paymentMethods: methods,
      };

      // Store the updated user data
      await storage.setItem("user", JSON.stringify(updatedUser));

      // Update state
      setState((prev) => ({
        ...prev,
        user: updatedUser,
      }));

      console.log("Payment methods updated successfully:", methods);
    } catch (error) {
      console.error("Error updating payment methods:", error);
      throw error;
    }
  };

  const updatePaymentMethod = async (hasPaymentMethod: boolean) => {
    try {
      if (!state.user) {
        throw new Error("No authenticated user");
      }

      // Update user with new payment method
      const updatedUser = {
        ...state.user,
        paymentMethods: hasPaymentMethod
          ? [
              {
                id: "new_payment_method",
                last4: "XXXX",
                brand: "New",
                expiryMonth: "12",
                expiryYear: "2024",
                isDefault: true,
              },
            ]
          : [],
      };

      // Store the updated user data
      await storage.setItem("user", JSON.stringify(updatedUser));

      // Update state
      setState((prev) => ({
        ...prev,
        user: updatedUser,
      }));

      console.log(
        "Payment method updated successfully:",
        hasPaymentMethod ? "Added" : "Removed"
      );
    } catch (error) {
      console.error("Error updating payment method:", error);
      throw error;
    }
  };

  const value = {
    ...state,
    signIn,
    signUp,
    signOut,
    signInWithGoogle,
    signInWithFacebook,
    updateProfile,
    updateFirstTimeProfile,
    updateMembership,
    updatePaymentMethods,
    updatePaymentMethod,
    isAuthenticated: !!state.token,
  };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
}

export function useAuth() {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
}
