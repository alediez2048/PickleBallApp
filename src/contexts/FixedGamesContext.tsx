// Context for Fixed Games CRUD operations
import React, { createContext, useContext, useState, useCallback } from "react";
import {
  getFixedGames,
  getFixedGame,
  createFixedGame,
  updateFixedGame,
  deleteFixedGame,
} from "@/services/fixedGamesService";
import { FixedGame, FixedGameInput } from "@/types/fixedGames";

interface FixedGamesContextProps {
  fixedGames: FixedGame[];
  loading: boolean;
  fetchFixedGames: () => Promise<void>;
  getFixedGame: (id: string) => Promise<FixedGame | null>;
  createFixedGame: (input: FixedGameInput) => Promise<FixedGame | null>;
  updateFixedGame: (
    id: string,
    input: Partial<FixedGameInput>
  ) => Promise<boolean>;
  deleteFixedGame: (id: string) => Promise<boolean>;
}

const FixedGamesContext = createContext<FixedGamesContextProps | undefined>(
  undefined
);

export const FixedGamesProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [fixedGames, setFixedGames] = useState<FixedGame[]>([]);
  const [loading, setLoading] = useState(false);

  const fetchFixedGames = useCallback(async () => {
    setLoading(true);
    try {
      const data = await getFixedGames();
      setFixedGames(data);
    } finally {
      setLoading(false);
    }
  }, []);

  const getFixedGameById = async (id: string) => {
    return await getFixedGame(id);
  };

  const create = async (input: FixedGameInput) => {
    setLoading(true);
    try {
      const created = await createFixedGame(input);
      if (created) await fetchFixedGames();
      return created;
    } finally {
      setLoading(false);
    }
  };

  const update = async (id: string, input: Partial<FixedGameInput>) => {
    setLoading(true);
    try {
      const ok = await updateFixedGame(id, input);
      if (ok) await fetchFixedGames();
      return ok;
    } finally {
      setLoading(false);
    }
  };

  const remove = async (id: string) => {
    setLoading(true);
    try {
      const ok = await deleteFixedGame(id);
      if (ok) await fetchFixedGames();
      return ok;
    } finally {
      setLoading(false);
    }
  };

  return (
    <FixedGamesContext.Provider
      value={{
        fixedGames,
        loading,
        fetchFixedGames,
        getFixedGame: getFixedGameById,
        createFixedGame: create,
        updateFixedGame: update,
        deleteFixedGame: remove,
      }}
    >
      {children}
    </FixedGamesContext.Provider>
  );
};

export function useFixedGames() {
  const ctx = useContext(FixedGamesContext);
  if (!ctx)
    throw new Error("useFixedGames must be used within FixedGamesProvider");
  return ctx;
}
