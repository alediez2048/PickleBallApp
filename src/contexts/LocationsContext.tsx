import React, {
  createContext,
  useContext,
  useState,
  useCallback,
  useEffect,
  ReactNode,
} from "react";
import * as locationsService from "@/services/locationsService";
import { Location } from "@/types/locations";

interface LocationsContextProps {
  locations: Location[];
  loading: boolean;
  error: string | null;
  fetchLocations: () => Promise<void>;
  getLocation: (id: string) => Promise<Location | null>;
  createLocation: (
    data: Omit<Location, "id" | "created_at">
  ) => Promise<Location | null>;
  updateLocation: (id: string, updates: Partial<Location>) => Promise<boolean>;
  deleteLocation: (id: string) => Promise<boolean>;
}

const LocationsContext = createContext<LocationsContextProps | undefined>(
  undefined
);

export const LocationsProvider = ({ children }: { children: ReactNode }) => {
  const [locations, setLocations] = useState<Location[]>([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  const fetchLocations = useCallback(async () => {
    setLoading(true);
    setError(null);
    const { data, error } = await locationsService.listLocations();
    if (error) setError(error.message);
    setLocations(data || []);
    setLoading(false);
  }, []);

  const getLocation = useCallback(async (id: string) => {
    const { data, error } = await locationsService.getLocation(id);
    if (error) setError(error.message);
    return data || null;
  }, []);

  const createLocation = useCallback(
    async (data: Omit<Location, "id" | "created_at">) => {
      const { data: created, error } =
        await locationsService.createLocation(data);
      if (error) {
        setError(error.message);
        return null;
      }
      await fetchLocations();
      return created?.[0] || null;
    },
    [fetchLocations]
  );

  const updateLocation = useCallback(
    async (id: string, updates: Partial<Location>) => {
      const { error } = await locationsService.updateLocation(id, updates);
      if (error) {
        setError(error.message);
        return false;
      }
      await fetchLocations();
      return true;
    },
    [fetchLocations]
  );

  const deleteLocation = useCallback(
    async (id: string) => {
      const { error } = await locationsService.deleteLocation(id);
      if (error) {
        setError(error.message);
        return false;
      }
      await fetchLocations();
      return true;
    },
    [fetchLocations]
  );

  useEffect(() => {
    fetchLocations();
  }, [fetchLocations]);

  return (
    <LocationsContext.Provider
      value={{
        locations,
        loading,
        error,
        fetchLocations,
        getLocation,
        createLocation,
        updateLocation,
        deleteLocation,
      }}
    >
      {children}
    </LocationsContext.Provider>
  );
};

export const useLocations = () => {
  const context = useContext(LocationsContext);
  if (!context)
    throw new Error("useLocations must be used within a LocationsProvider");
  return context;
};
