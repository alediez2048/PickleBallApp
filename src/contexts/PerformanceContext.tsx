import React, { createContext, useContext, useCallback, useState } from 'react';
import type { PerformanceMetrics } from '@/hooks/usePerformanceMonitor';

interface PerformanceContextType {
  metrics: Record<string, PerformanceMetrics[]>;
  addMetrics: (componentName: string, metrics: PerformanceMetrics) => void;
  getAverageMetrics: (componentName: string) => PerformanceMetrics | null;
  clearMetrics: (componentName?: string) => void;
}

const PerformanceContext = createContext<PerformanceContextType | undefined>(undefined);

export function PerformanceProvider({ children }: { children: React.ReactNode }) {
  const [metrics, setMetrics] = useState<Record<string, PerformanceMetrics[]>>({});

  const addMetrics = useCallback((componentName: string, newMetrics: PerformanceMetrics) => {
    setMetrics(prev => ({
      ...prev,
      [componentName]: [...(prev[componentName] || []), newMetrics],
    }));
  }, []);

  const getAverageMetrics = useCallback((componentName: string): PerformanceMetrics | null => {
    const componentMetrics = metrics[componentName];
    if (!componentMetrics?.length) return null;

    return {
      mountTime: componentMetrics.reduce((acc, m) => acc + m.mountTime, 0) / componentMetrics.length,
      renderTime: componentMetrics.reduce((acc, m) => acc + m.renderTime, 0) / componentMetrics.length,
      interactionTime: componentMetrics.reduce((acc, m) => acc + m.interactionTime, 0) / componentMetrics.length,
    };
  }, [metrics]);

  const clearMetrics = useCallback((componentName?: string) => {
    if (componentName) {
      setMetrics(prev => {
        const { [componentName]: _, ...rest } = prev;
        return rest;
      });
    } else {
      setMetrics({});
    }
  }, []);

  return (
    <PerformanceContext.Provider
      value={{
        metrics,
        addMetrics,
        getAverageMetrics,
        clearMetrics,
      }}
    >
      {children}
    </PerformanceContext.Provider>
  );
}

export function usePerformance() {
  const context = useContext(PerformanceContext);
  if (context === undefined) {
    throw new Error('usePerformance must be used within a PerformanceProvider');
  }
  return context;
} 