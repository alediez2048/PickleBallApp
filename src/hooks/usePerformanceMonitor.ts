import { useRef, useEffect } from 'react';
import { InteractionManager } from 'react-native';

export interface PerformanceMetrics {
  mountTime: number;
  renderTime: number;
  interactionTime: number;
}

interface PerformanceMonitorOptions {
  componentName: string;
  onMetricsCollected?: (metrics: PerformanceMetrics) => void;
}

export function usePerformanceMonitor({ componentName, onMetricsCollected }: PerformanceMonitorOptions) {
  const startTimeRef = useRef<number>(0);
  const mountTimeRef = useRef<number>(0);
  const renderTimeRef = useRef<number>(0);

  useEffect(() => {
    startTimeRef.current = performance.now();

    const interaction = InteractionManager.createInteractionHandle();

    return () => {
      const endTime = performance.now();
      const metrics: PerformanceMetrics = {
        mountTime: mountTimeRef.current - startTimeRef.current,
        renderTime: renderTimeRef.current - startTimeRef.current,
        interactionTime: endTime - startTimeRef.current,
      };

      console.debug(`[Performance] ${componentName}:`, metrics);
      onMetricsCollected?.(metrics);

      InteractionManager.clearInteractionHandle(interaction);
    };
  }, [componentName, onMetricsCollected]);

  useEffect(() => {
    mountTimeRef.current = performance.now();
  }, []);

  useEffect(() => {
    renderTimeRef.current = performance.now();
  });

  return null;
} 