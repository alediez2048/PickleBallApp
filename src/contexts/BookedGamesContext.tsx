import React, { createContext, useContext, useState, useEffect } from "react";
import { useAuth } from "@/contexts/AuthContext";
import {
  createBookedGame as createBookedGameService,
  listBookedGames as listBookedGamesService,
  updateBookedGame as updateBookedGameService,
} from "@services/bookedGamesService";
import { getGame, updateGame } from "@services/gamesService";

import { BookedGame } from "@/types/bookedGames";
import { User } from "@/types/games";

interface BookedGamesContextType {
  bookedGames: BookedGame[];
  isLoading: boolean;
  error: Error | null;
  refreshBookedGames: () => Promise<void>;
  addBookedGame: (game: Omit<BookedGame, "id">) => Promise<BookedGame | null>;
  cancelBooking: (gameId: string) => Promise<void>;
  clearAllGames: () => Promise<void>;
  listBookedGamesForUser: () => Promise<BookedGame[]>;
}

const BookedGamesContext = createContext<BookedGamesContextType | undefined>(
  undefined
);

export const BookedGamesProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [bookedGames, setBookedGames] = useState<BookedGame[]>([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<Error | null>(null);
  const { user } = useAuth();

  const refreshBookedGames = async () => {
    if (!user?.email) return;
    setIsLoading(true);
    setError(null);
    try {
      const { data, error: fetchError } = await listBookedGamesService();
      if (fetchError) throw fetchError;
      setBookedGames(data || []);
    } catch (err) {
      setError(
        err instanceof Error ? err : new Error("Failed to fetch booked games")
      );
    } finally {
      setIsLoading(false);
    }
  };

  const addBookedGame = async (game: Omit<BookedGame, "id">) => {
    if (!user?.email) throw new Error("User not authenticated");
    try {
      // Fetch the current game details first
      const currentGameData = await getGame(game.game_id);
      if (!currentGameData) throw new Error("Game not found");

      // Prepare the new player object using User type
      const newPlayer: User = {
        id: user.id,
        email: user.email,
        name: user.display_name || user.name || "",
        skill_level: currentGameData.skill_level,
        phone_number: user.phone_number || "",
      };

      // Prepare the updated players array
      const updatedPlayers: User[] = [
        ...(currentGameData.players || []),
        newPlayer,
      ];

      // Register the booked game
      const bookedGame = await createBookedGameService({ ...game });
      if (bookedGame) {
        await updateGame(bookedGame.game_id, {
          registered_count: updatedPlayers.length,
          players: updatedPlayers,
        });

        setBookedGames((prev) => [bookedGame, ...prev]);
        return bookedGame;
      }
      return null;
    } catch (err) {
      console.error("Error adding booked game:", err);
      throw err instanceof Error ? err : new Error("Failed to book game");
    }
  };

  const cancelBooking = async (gameId: string) => {
    if (!user?.email) throw new Error("User not authenticated");
    try {
      await updateBookedGameService(gameId, { status: "cancelled" });
      setBookedGames((prev) =>
        prev.map((game) =>
          game.id === gameId ? { ...game, status: "cancelled" as const } : game
        )
      );
    } catch (err) {
      throw err instanceof Error ? err : new Error("Failed to cancel booking");
    }
  };

  const clearAllGames = async () => {
    if (!user?.email) return;
    setIsLoading(true);
    try {
      setBookedGames([]);
    } catch (err) {
      throw err instanceof Error ? err : new Error("Failed to clear games");
    } finally {
      setIsLoading(false);
    }
  };

  /**
   * Fetch upcoming booked games for the current user in the next 7 days
   */
  const listBookedGamesForUser = async () => {
    if (!user?.id) return [];
    const today = new Date();
    const startDate = today.toISOString().split("T")[0];
    const endDateObj = new Date(today);
    endDateObj.setDate(today.getDate() + 7);
    const endDate = endDateObj.toISOString().split("T")[0];
    const { data, error } = await listBookedGamesService({
      userId: user.id,
      dateRange: { startDate, endDate },
      status: "upcoming",
    });
    if (error) {
      console.error("Error fetching upcoming booked games for user:", error);
      return [];
    }
    return data || [];
  };

  useEffect(() => {
    refreshBookedGames();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [user?.email]);

  return (
    <BookedGamesContext.Provider
      value={{
        bookedGames,
        isLoading,
        error,
        refreshBookedGames,
        addBookedGame,
        cancelBooking,
        clearAllGames,
        listBookedGamesForUser,
      }}
    >
      {children}
    </BookedGamesContext.Provider>
  );
};

export function useBookedGames() {
  const context = useContext(BookedGamesContext);
  if (context === undefined) {
    throw new Error("useBookedGames must be used within a BookedGamesProvider");
  }
  return context;
}
