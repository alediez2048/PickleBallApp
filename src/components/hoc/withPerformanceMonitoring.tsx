import React from 'react';
import { usePerformanceMonitor } from '@/hooks/usePerformanceMonitor';
import { usePerformance } from '@/contexts/PerformanceContext';

export function withPerformanceMonitoring<P extends object>(
  WrappedComponent: React.ComponentType<P>,
  options: { componentName?: string } = {}
) {
  const displayName = options.componentName || WrappedComponent.displayName || WrappedComponent.name || 'Component';

  function WithPerformanceMonitoring(props: P) {
    const { addMetrics } = usePerformance();

    usePerformanceMonitor({
      componentName: displayName,
      onMetricsCollected: (metrics) => {
        addMetrics(displayName, metrics);
      },
    });

    return <WrappedComponent {...props} />;
  }

  WithPerformanceMonitoring.displayName = `withPerformanceMonitoring(${displayName})`;

  return WithPerformanceMonitoring;
} 