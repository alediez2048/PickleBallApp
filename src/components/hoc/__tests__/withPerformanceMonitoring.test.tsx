import React from 'react';
import { Text } from 'react-native';
import renderer from 'react-test-renderer';
import { withPerformanceMonitoring } from '../withPerformanceMonitoring';
import { PerformanceProvider } from '@/contexts/PerformanceContext';
import { usePerformanceMonitor } from '@/hooks/usePerformanceMonitor';

// Mock the usePerformanceMonitor hook
jest.mock('@/hooks/usePerformanceMonitor', () => ({
  usePerformanceMonitor: jest.fn(),
}));

// Mock the PerformanceProvider context
jest.mock('@/contexts/PerformanceContext', () => ({
  usePerformance: () => ({
    addMetrics: jest.fn(),
  }),
  PerformanceProvider: ({ children }: { children: React.ReactNode }) => <>{children}</>,
}));

describe('withPerformanceMonitoring', () => {
  const TestComponent = ({ text }: { text: string }) => <Text>{text}</Text>;
  TestComponent.displayName = 'TestComponent';

  const WrappedComponent = withPerformanceMonitoring(TestComponent);

  beforeEach(() => {
    (usePerformanceMonitor as jest.Mock).mockClear();
  });

  it('renders the wrapped component with its props', () => {
    const tree = renderer
      .create(
        <PerformanceProvider>
          <WrappedComponent text="Hello" />
        </PerformanceProvider>
      )
      .toJSON();
    
    expect(tree).toMatchSnapshot('component with performance monitoring');
  });

  it('sets correct display name', () => {
    expect(WrappedComponent.displayName).toBe('withPerformanceMonitoring(TestComponent)');
  });

  it('uses custom component name when provided', () => {
    const CustomNameComponent = withPerformanceMonitoring(TestComponent, {
      componentName: 'CustomName',
    });

    expect(CustomNameComponent.displayName).toBe('withPerformanceMonitoring(CustomName)');
  });

  it('initializes performance monitoring with correct parameters', () => {
    renderer.create(
      <PerformanceProvider>
        <WrappedComponent text="Hello" />
      </PerformanceProvider>
    );

    expect(usePerformanceMonitor).toHaveBeenCalledWith({
      componentName: 'TestComponent',
      onMetricsCollected: expect.any(Function),
    });
  });

  it('handles components without display name', () => {
    const AnonymousComponent = () => <Text>Anonymous</Text>;
    const WrappedAnonymous = withPerformanceMonitoring(AnonymousComponent);

    const tree = renderer
      .create(
        <PerformanceProvider>
          <WrappedAnonymous />
        </PerformanceProvider>
      )
      .toJSON();
    
    expect(tree).toMatchSnapshot('anonymous component with performance monitoring');
    expect(WrappedAnonymous.displayName).toMatch(/withPerformanceMonitoring\(.+\)/);
  });
}); 