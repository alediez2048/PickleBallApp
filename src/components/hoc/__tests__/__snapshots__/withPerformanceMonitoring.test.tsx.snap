// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withPerformanceMonitoring handles components without display name: anonymous component with performance monitoring 1`] = `null`;

exports[`withPerformanceMonitoring renders the wrapped component with its props: component with performance monitoring 1`] = `null`;
